docs/content/functions/test.md

```diff
@@ -69,3 +69,57 @@ template: <arg>:1:3: executing "<arg>" at <fail>: error calling fail: template g
 $ gomplate -i '{{ test.Fail "something is wrong!" }}'
 template: <arg>:1:7: executing "<arg>" at <test.Fail>: error calling Fail: template generation failed: something is wrong!
 ```
+
+## `test.Required`
+
+**Alias:** `required`
+
+Passes through the given value, if it's non-empty, non-zero, and non-`nil`.
+Otherwise, exits and prints a given error message so the user can adjust as
+necessary.
+
+This is particularly useful for cases where templates require user-provided
+data (such as datasources or environment variables), and rendering can not
+continue correctly.
+
+This was inspired by [Helm's `required` function](https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#know-your-template-functions),
+but has slightly different behaviour. Notably, gomplate will always fail in
+cases where a referenced _key_ is missing, and this function will have no
+effect.
+
+### Usage
+```go
+test.Required [message] value 
+```
+
+```go
+value | test.Required [message]  
+```
+
+### Arguments
+
+| name | description |
+|------|-------------|
+| `message` | _(optional)_ The optional message to provide when the required value is not provided |
+| `value` | _(required)_ The required value |
+
+### Examples
+
+```console
+$ FOO=foobar gomplate -i '{{ getenv "FOO" | required "Missing FOO environment variable!" }}'
+foobar
+$ FOO= gomplate -i '{{ getenv "FOO" | required "Missing FOO environment variable!" }}'
+error: Missing FOO environment variable!
+```
+```console
+$ cat <<EOF> config.yaml
+defined: a value
+empty: ""
+EOF
+$ gomplate -d config=config.yaml -i '{{ (ds "config").defined | required "The `config` datasource must have a value defined for `defined`" }}'
+a value
+$ gomplate -d config=config.yaml -i '{{ (ds "config").empty | required "The `config` datasource must have a value defined for `empty`" }}'
+template: <arg>:1:25: executing "<arg>" at <required "The `confi...>: error calling required: The `config` datasource must have a value defined for `empty`
+$ gomplate -d config=config.yaml -i '{{ (ds "config").bogus | required "The `config` datasource must have a value defined for `bogus`" }}'
+template: <arg>:1:7: executing "<arg>" at <"config">: map has no entry for key "bogus"
```

Comment:
Is there a way we could do something like:
```
{{ (ds "config") | get "bogus" | required "The `config` datasource must have a value defined for `bogus`"
```

where `get` works like `getenv` and does not error?

Reply:
haven't tried...

Reply:
ah, that might work. There is the `index` built-in, but it doesn't pipeline so well:

```console
$  bin/gomplate -i '{{ $j := (json `{"foo":"bar"}`)}}{{ index $j "baz" | required "nope!" }}'
template: <arg>:1:53: executing "<arg>" at <required "nope!">: error calling required: nope!
```

_if only_ `index` took arguments the other way around, it'd be a simple case of `{{ $j | index "baz" | required "nope!" }}`.

I could write an alternate for `index` that takes arguments in the opposite order for easier pipelining, but it may not be _that_ necessary...